barretenberg/cpp/src/barretenberg/vm/avm/trace/execution.cpp

```diff
@@ -258,585 +294,637 @@ bool Execution::verify(AvmFlavor::VerificationKey vk, HonkProof const& proof)
  *
- * @param instructions A vector of the instructions to be executed.
- * @param calldata expressed as a vector of finite field elements.
- * @param public_inputs expressed as a vector of finite field elements.
+ * @param public_inputs - to constrain execution inputs & results against
+ * @param returndata - to add to for each enqueued call
+ * @param execution_hints - to inform execution
+ * @param apply_end_gas_assertions - should we apply assertions that public input's end gas is right?
  * @return The trace as a vector of Row.
  */
-std::vector<Row> Execution::gen_trace(std::vector<FF> const& calldata,
-                                      AvmPublicInputs const& public_inputs,
+std::vector<Row> Execution::gen_trace(AvmPublicInputs const& public_inputs,
                                       std::vector<FF>& returndata,
-                                      ExecutionHints const& execution_hints)
+                                      ExecutionHints const& execution_hints,
+                                      bool apply_end_gas_assertions)
 
 {
     vinfo("------- GENERATING TRACE -------");
     // TODO(https://github.com/AztecProtocol/aztec-packages/issues/6718): construction of the public input columns
     // should be done in the kernel - this is stubbed and underconstrained
     // VmPublicInputs public_inputs = avm_trace::convert_public_inputs(public_inputs_vec);
-    uint32_t start_side_effect_counter =
-        0; // What to do here???
-           // !public_inputs_vec.empty() ?
-           // static_cast<uint32_t>(public_inputs_vec[START_SIDE_EFFECT_COUNTER_PCPI_OFFSET])
-           //                            : 0;
-           //
+    uint32_t start_side_effect_counter = 0;
+    // Temporary until we get proper nested call handling
+    std::vector<FF> calldata;
+    for (const auto& enqueued_call_hints : execution_hints.enqueued_call_hints) {
+        calldata.insert(calldata.end(), enqueued_call_hints.calldata.begin(), enqueued_call_hints.calldata.end());
+    }
     AvmTraceBuilder trace_builder =
         Execution::trace_builder_constructor(public_inputs, execution_hints, start_side_effect_counter, calldata);
 
-    std::vector<PublicCallRequest> public_call_requests;
-    for (const auto& setup_requests : public_inputs.public_setup_call_requests) {
-        if (setup_requests.contract_address != 0) {
-            public_call_requests.push_back(setup_requests);
-        }
-    }
-    for (const auto& app_requests : public_inputs.public_app_logic_call_requests) {
-        if (app_requests.contract_address != 0) {
-            public_call_requests.push_back(app_requests);
-        }
+    const auto setup_call_requests = non_empty_call_requests(public_inputs.public_setup_call_requests);
+    const auto app_logic_call_requests = non_empty_call_requests(public_inputs.public_app_logic_call_requests);
+    std::vector<PublicCallRequest> teardown_call_requests;
+    if (!public_inputs.public_teardown_call_request.is_empty()) {
```

Comment:
Do we have to keep this if condition? If so, a comment why we need it and why do we have different code flow in setting teardown_call_requests would be great.

Reply:
It's because teardown isn't an array. It's a single call request as commented here. But you're right, I'll add more.